### ISSUE TITLE
[Android] APP is not Document Provider, cannot interact with configs/shaders/etc

### ISSUE THREAD

**Issue by Dogway** [libretro/RetroArch#15309]
## Description

I opened a [thread](https://forums.libretro.com/t/retroarch-not-saving-global-default-config/39612) a half a year ago with this same issue but just reading Dolphin's last blog entry I got the proper term.

The retroarch APP by design seems to not be Document Provider, that is; *"This means you can use Dolphin to directly copy files into and out of its  per-app directory  on the latest versions of Android."*
So after installing Retroarch on GoogleTV I cannot load my configs and settings from the desktop version because I have not access to those folders (*Android/data/RetroArch*). Or add custom shaders/bezels, etc.

### Expected behavior

I expect the APP to be Document Provider like Dolphin, Kodi, etc, and be configurable via a built-in browser.

### Actual behavior

Cannot add or remove files (shaders, bezels, core configs, core options, all types of overrides, custom UI, etc)
Cannot save default options, retroarch always starts "clean".

### Version/Commit

- RetroArch: RetroArch_v1.13.0_armeabi-v7a_32bit

### Environment information

- OS: Chromecast HD's GoogleTV

**Comment by hizzlekizzle** [comment]
That's what this is, right? https://github.com/libretro/RetroArch/blob/master/pkg/android/phoenix/src/com/retroarch/browser/provider/RetroDocumentsProvider.java

**Comment by Dogway** [comment]
I guess so. The problem and as explained in the Dolphin blog [entry](https://es.dolphin-emu.org/blog/2023/05/21/dolphin-progress-report-february-march-april-2023/?#50-18922-android-implement-document-provider-support-by-k0bin) is that newer Androids limits the access to protected folders so even a file browser with Document Provider support (Files?) won't work. This is explained further [here](https://developer.android.com/training/data-storage/shared/documents-files?#document-tree-access-restrictions).

**Comment by Dogway** [comment]
Chromecast is not powerful enough for Dolphin so I will install first RA and then Dolphin on my tablet and see if Files has access to it, looks like that's the file browser used in the screenshot for the blog entry.

AndroidTV doesn't support Files I think so a workaround would be needed (like Kodi) to inject the modified files.

Will report back.

**Comment by Dogway** [comment]
### Dolphin
Installed both APPs as latest nightlies. Dolphin stores its "program files" within:
`/storage/emulated/0/Android/data/org.dolphinemu.dolphinemu`

I have three file browsers on the tablet; Google Files, Samsung's own file browser and FileCommander.
Files and Samsung's file browser don't have access to `Android/data`, while FileCommander asks for a special access grant to a protected folder as shown in this screenshot.

[Screenshot](https://photos.app.goo.gl/uwiEt23ZeWScsrsN8)

After grant is conceded I proceed to access the folder with a following denied access. So something is not right there.

### RetroArch
Meanwhile with RetroArch it stores its files in three different locations:
```
/data/user/0/com.retroarch.aarch64 (default RA "program files" location)
/storage/emulated/0/Android/data/com.retroarch.aarch64 (?)
/storage/emulated/0/Retroarch (user files)
```

I repeat the FileCommander procedure and I'm granted access to the `/storage/emulated/0/Android/data` path this time without issues, I can edit the main default retroarch.config. `/data/user/0/` cannot be granted any access

So while cumbersome and unideal a workaround is possible where I can edit default's retroarch.config to assign all the "program files" folders to `/storage/emulated/0/` path and start customizing files. This won't move the files so you will have to repopulate the default files bundled with the APP (about 300Mb), **only then** you can download resources with the online updater.

Translating this to AndroidTV is not possible as FileCommander for some reason fails to load the option to grant access to special folders. So another file browser (which?) should be used or as in the Kodi example a built-in browser is implemented, my suggestion.

**Comment by Dogway** [comment]
I had time to take a deeper look into this from Dolphin's POV.
They provide a way to access their Data folder from within Dolphin's APP:
`/storage/emulated/0/Android/data/org.dolphinemu.dolphinemu`
via Settings->config->userdata->Open User Data Folder

With this a browser opens Dolphin's config folder.

It happens to be that all Android OS have a built-in file browser usually hidden by the manufacturer, and it's called "Files" or "Archivos" in Spanish and probably different for different languages.

You can either use a third party APP with Document Provider enabled such as [ZArchiver](https://play.google.com/store/apps/details?id=ru.zdevs.zarchiver&hl=en&gl=US), find a mean to open Android's built-in Files APP (not to confuse with Google's Files APP), or as I did download a third party [shortcut](https://play.google.com/store/apps/details?id=com.marc.files) to Files.

So to sum it up, unless you rely on third party APPs it's very difficult if not impossible to access the config folders for RetroArch.
So my suggestion is to at least if not including a file browser within RA, to provide a "Files" APP shortcut or shortcuts within RA to the different RA config folders.
